scripts/gen-exprs.py

```diff
@@ -0,0 +1,940 @@
+
+
+class RefFunc(Expression):
+    func = Name()
+
+
+class RefEq(Expression):
+    left = Child()
+    right = Child()
+
+
+class Try(Expression):
+    body = Child()
+    catchBody = Child()
+
+    methods = [
+        'void finalize();',
+        'void finalize(Type type_);',
+    ]
+
+
+class Throw(Expression):
+    event = Name()
+    operands = ChildList()
+
+
+class Rethrow(Expression):
+    exnref = Child()
+
+
+class BrOnExn(Expression):
+    __constructor_body__ = 'type = Type::unreachable;'
+
+    name = ScopeNameUse()
+    event = Name()
+    exnref = Child()
+
+    """
+    This is duplicate info of param types stored in Event, but this is required
+    for us to know the type of the value sent to the target block.
+    """
+    sent = Type()
+
+
+class TupleMake(Expression):
+    operands = ChildList()
+
+
+class TupleExtract(Expression):
+    tuple = Child()
+    index = Index()
+
+
+class I31New(Expression):
+    value = Child()
+
+
+class I31Get(Expression):
+    i31 = Child()
+    signed_ = Bool()
+
+
+class RefTest(Expression):
+    pass
+
+
+class RefCast(Expression):
+    pass
+
+
+class BrOnCast(Expression):
+    pass
+
+
+class RttCanon(Expression):
+    pass
+
+
+class RttSub(Expression):
+    pass
+
+
+class StructNew(Expression):
+    pass
+
+
+class StructGet(Expression):
+    pass
+
+
+class StructSet(Expression):
+    pass
+
+
+class ArrayNew(Expression):
+    pass
+
+
+class ArrayGet(Expression):
+    pass
+
+
+class ArraySet(Expression):
+    pass
+
+
+class ArrayLen(Expression):
+    pass
+
+
+###############
+# Boilerplate
+###############
+
+
+COPYRIGHT = f"""\
+/*
+ * Copyright {datetime.datetime.now().year} WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+"""
+
+NOTICE = f"""\
+//=============================================================================
+// This is an AUTOGENERATED file, even though it looks human-readable! Do not
+// edit it by hand, instead edit the generator, {os.path.basename(__file__)}
+//
+// You can and should treat it like human-written code in all other ways,
+// though, like reviewing it in a PR, etc.
+//=============================================================================
+"""
+
+BOILERPLATE = COPYRIGHT + '\n' + NOTICE
+
+
+##############
+# Processing
+##############
+
+
+def is_subclass_of(x, y):
+    """Whether x is a subclass of y."""
+
+    return y in getattr(x, '__bases__', {})
+
+
+def is_a(x, cls):
```

Comment:
This confused me at first because in LLVM's RTTI system, `isa` corresponds to `is_subclass_of` here. We could call this `is_instance` or `is_instance_of` instead?

Reply:
Sounds good, will change to `is_instance`.